demos/bounded_heat/bounded_heat.py.rst

```diff
@@ -0,0 +1,202 @@
+
+    rhs = expand_derivatives(diff(uexact, t)) - div(grad(uexact))
+
+Note that the exact solution is uniformly positive in space and time. Using a manufactured 
+solution, one would typically project the exact solution at time :math:`t = 0` onto the 
+approximation space and use the projection as the initial condition. In this case, the 
+projection does not satisfy the lower bound of :math:`0`. We instead solve a variational inequality 
+to find a bounds-preserving initial condition: ::
+
+    v = TestFunction(V)
+    u = Function(V)
+
+    G = inner(u - uexact, v) * dx
+
+    nlvp = NonlinearVariationalProblem(G, u)
+    nlvs = NonlinearVariationalSolver(nlvp, solver_parameters=vi_params)
+
+    upper = Function(V)
+    lower = Function(V)
+
+    with upper.dat.vec as upper_vec:
+        upper_vec.set(np.inf)
+
+    with lower.dat.vec as lower_vec:
+        lower_vec.set(0.0)
+
+    nlvs.solve(bounds=(lower, upper))
+
+    u_c = u.copy(deepcopy=True)
+
+``u`` and ``u_c`` now hold a bounds-constrained approximation to the exact solution 
+at :math:`t = 0`.
+
+We now construct semidiscrete variational problems for both the constrained and unconstrained 
+approximations using UFL notation and the ``Dt`` operator from Irksome: ::
+
+    v = TestFunction(V)
+
+    F = (inner(Dt(u), v) * dx + inner(grad(u), grad(v)) * dx - inner(rhs, v) * dx)
+
+    v_c = TestFunction(V)
+
+    F_c = (inner(Dt(u_c), v_c) * dx + inner(grad(u_c), grad(v_c)) * dx - inner(rhs, v_c) * dx)
+
+We use exact boundary conditions in both cases: ::
+
+    bc = DirichletBC(V, uexact, "on_boundary")
+
+For the unconstrained approximation, we configure the :class:`.TimeStepper` in a 
+familiar way: ::
+
+    stepper = TimeStepper(F, butcher_tableau, t, dt, u, bcs=bc, solver_parameters=lu_params)
+
+We will enforce nonnegativity when finding the constrained approximation. We now set up the keyword database to 
+configure an instance of :class:`.TimeStepper` for this task. We first specify, using the 
+keyword ``stage_type``, that we wish to use a stage-value formulation of the underlying collocation 
+method. The keyword ``basis_type`` then allows us to change the basis of the collocation 
+polynomial to the Bernstein basis. Having done this, we must specify a solver which is able to handle bounds 
+constraints. In this example we solve a variational inequality using ``vinewtonrsls`` by passing ``vi_params`` 
+as ``solver_parameters`` to the :class:`.TimeStepper`: ::
+
+    kwargs_c = {"stage_type": "value",
+                "basis_type": 'Bernstein',
+                "solver_parameters": vi_params
+            }
+
+    stepper_c = TimeStepper(F_c, butcher_tableau, t, dt, u_c, bcs=bc, **kwargs_c)
+
+We set the bounds as follows: ::
+
+    lb = Function(V)
+    lb.assign(0)
+
+    ub = None
+
+    bounds = ('stage', lb, ub)
+
+Passing the :meth:`~.TimeStepper.advance` method ``bounds`` will enforce the 
```

Comment:
This is not (quite) right.  There is no `TimeStepper` class since that is a convenience function that takes the keyword arguments and constructs a class of some type.  In fact, `advance` only supports the `bounds` argument if we are in stage value form (Bernstein or standard is fine).